Source/warningreporting/WarningReportingUnit.cpp

```diff
@@ -234,120 +187,87 @@ namespace WarningReporting {
         }
 
-        for (IWarningReportingUnit::IWarningReportingControl* warningControl : m_Categories) {
-            Settings::const_iterator index = m_EnabledCategories.begin(); 
-            while (index != m_EnabledCategories.end()) {
-                const Setting& setting = *index;
-
-                if ( setting.Category() == warningControl->Category() ) {
-                    if( setting.Enabled() != warningControl->Enabled() ) {
-                        warningControl->Enabled(setting.Enabled()); 
-                    }
-                    warningControl->Configure(setting.Configuration());
-                    break; // HPL todo: you might to add this also on TraceControl
+        for (auto& setting : _enabledCategories) {
+            auto category = _categories.find(setting.first);
+
+            if (category != _categories.end()) {
+                
+                if (category->second->Enabled() != setting.second.Enabled()) {
+                    category->second->Enabled(setting.second.Enabled());
                 }
 
-                index++;
+                category->second->Configure(setting.second.Configuration());
+                category->second->Exclude(setting.second.Excluded());
             }
         }
     }
 
-    bool WarningReportingUnit::IsDefaultCategory(const string& category, bool& enabled, string& config) const
+    void WarningReportingUnit::FetchCategoryInformation(const string& category, bool& outIsDefaultCategory, bool& outIsEnabled, string& outExcluded, string& outConfiguration) const
     {
+        Core::SafeSyncType<Core::CriticalSection> guard(_adminLock);
 
-        bool isDefaultCategory = false;
+        outIsDefaultCategory = true;
+        auto setting = _enabledCategories.find(category);
 
-        Settings::const_iterator index = m_EnabledCategories.begin(); 
-        while (index != m_EnabledCategories.end()) {
-            const Setting& setting = *index;
-
-            if ( setting.Category() == category) {
-                isDefaultCategory = true;
-                enabled = setting.Enabled();
-                config = setting.Configuration();
-                break; // HPL todo: also in tracecontrol you probably want to stop the loop once found
-            }
-            index++; 
+        if (setting != _enabledCategories.end()) {
+            outIsDefaultCategory = true;
+            outIsEnabled = setting->second.Enabled();
+            outExcluded = setting->second.Excluded();
+            outConfiguration = setting->second.Configuration();
         }
-
-        return isDefaultCategory;
     }
 
     void WarningReportingUnit::ReportWarningEvent(const char identifier[], const char file[], const uint32_t lineNumber, const char className[], const IWarningEvent& information)
     {
 
         const char* fileName(Core::FileNameOnly(file));
 
-        m_Admin.Lock();
+        Core::SafeSyncType<Core::CriticalSection> guard(_adminLock);
```

Comment:
As mentioned above we have a rule to always use Lock and Unlock to prevent the lock being taken too long. Call me if you want the details, nice for a debate over a bear. Only exception is onelines for which SafeSyncType can be used